src/models/pairwise_union_reach_surface.py

```diff
@@ -0,0 +1,217 @@
+
+    r_1(imp_1)+r_2(imp_2)+...+r_p(imp_p) -
+      1/2 sum_{j=1..p} sum_{k=1..p}
+      (a_{j,k} * {r_j(imp_j)r_k(imp_k)} / {max\{m_j,m_k}})
+
+    where:
+          p is the number of publishers
+          a_i_j are the model parameters we want to estimate
+          imp_i is the numnber of impressions for ith publisher
+          r_j(imp_j) is the reach of the jth publisher for imp_j impressions.
+
+
+    """
+
+    def __init__(
+        self, reach_curves: Iterable[ReachCurve], reach_points: Iterable[ReachPoint]
+    ):
+        """Constructor for PaiwiseUnionReachSurface.
+
+        Args:
+          reach_curves: A list of ReachCurves to be used in model fitting and reach
+            prediction.
+          reach_points: A list of ReachPoints to which the model is to be fit. This
+            is parallel to the reach_curves list. The reach point at ith poisition
+            is drawn from ith reach curve.
+        """
+
+        self._reach_curves = copy.deepcopy(reach_curves)
+        self._n = len(reach_points)
+        self._p = len(reach_points[0].impressions)
+        super().__init__(data=reach_points, max_reach=0)
+
+    def _fit(self) -> None:
+        z, alpha = self.get_z_and_alpha()
+        self._a = self.solve_a_given_z_and_alpha(z, alpha)
+
+    def get_reach_vector(self, impressions: Iterable[int]) -> Iterable[int]:
+        """Calculates single publisher reaches for a given impression vector.
+
+        Args:
+          impressions: A list of impressions per publisher. prediction.
+
+        Returns:
+          Reach for each publisher.
+        """
+        return [
+            reach_curve.by_impressions(impression).reach()
+            for reach_curve, impression in zip(self._reach_curves, impressions)
+        ]
+
+    def by_impressions(
+        self, impressions: Iterable[int], max_frequency: int = 1
+    ) -> ReachPoint:
+
+        reach_vector = self.get_reach_vector(impressions)
+        reach_sum = sum(reach_vector)
+        overlap = sum(
+            [
+                (self._a[i * self._p + j] * reach_vector[i] * reach_vector[j])
+                / (
+                    max(
+                        self._reach_curves[i].max_reach, self._reach_curves[j].max_reach
+                    )
+                    * 2
+                )
+                for i in range(self._p)
+                for j in range(self._p)
+            ]
+        )
+
+        return ReachPoint(impressions, [reach_sum - overlap])
+
+    def get_inequality_constraints(self):
+        """Returns the inequality constraint for the Quadratic Programming solver.
+
+        3 types of inequality constraints are stacked in matrix G and vector h:
+
+          G1: a_i_j >= 0 for all i,j -> all parameters are non negative
+          G2: sum_{j=1..p} a_i_j <= 1 for all i -> row sums are no more than one
+          G3: sum_{i=1..p} a_i_j <= 1 for all j -> col sums are no more than one
```

Comment:
Perhaps clarify the return: 
Returns (matrix G, vector h). Then, cvxopt.solvers recognizes the constraint Gx <= h, where x is the parameter vector.

Reply:
+1